### PR TITLE
[PersistentCollection] Fix compatibility with PHP7 return types

### PR DESCRIPTION
**Pull request by Maks3w**
Solves the issue about signature mismatch due missing return type in child class.

### CODE REVIEW

**Comment by Maks3w** [comment]
I've tested in local tuning CustomCollectionsTest successfully. I didn't commit the changes due PHP5 vs PHP7 runs.

**Comment by alcaeus** [comment]
Great, I haven't even thought of that! With PHP 7.1 released, could you add support for nullable types and the void return type?

**Comment by Maks3w** [comment]
@alcaeus The code is based on Proxy code present at doctrine/commons. When that code become updated this could be done as well.

**Comment by alcaeus** [comment]
This was added to doctrine/common in 2.7.0: https://github.com/doctrine/common/pull/734

**Comment by Maks3w** [comment]
Then the code proposed on this PR include those changes. I just copy the code present on master yesterday.

**Comment by alcaeus** [comment]
Ok. I'll take a look later today.

File: lib/Doctrine/ODM/MongoDB/PersistentCollection/DefaultPersistentCollectionGenerator.php [hunk 1]
```diff
@@ -297,4 +298,65 @@ function (\ReflectionParameter $parameter) {
...
+     * @param \ReflectionParameter|null $parameter
+     *
+     * @return string
+     *
+     * @see \Doctrine\Common\Proxy\ProxyGenerator::formatType()
+     */
+    private function formatType(
+        \ReflectionType $type,
+        \ReflectionMethod $method,
+        \ReflectionParameter $parameter = null
```
Comment (Maks3w) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection/DefaultPersistentCollectionGenerator.php]: It's not my intention make something different from the code already present on doctrine commons.

If both snippets are equals then maintenance will be easy until a long term solution (DRY) become

Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection/DefaultPersistentCollectionGenerator.php]: I agree, we should try to keep it consistent.

File: lib/Doctrine/ODM/MongoDB/PersistentCollection/DefaultPersistentCollectionGenerator.php [hunk 2]
```diff
@@ -297,4 +298,65 @@ function (\ReflectionParameter $parameter) {
...
+        $nameLower = strtolower($name);
+        if ('self' === $nameLower) {
+            $name = $method->getDeclaringClass()->getName();
+        }
+        if ('parent' === $nameLower) {
+            $name = $method->getDeclaringClass()->getParentClass()->getName();
+        }
+        if ( ! $type->isBuiltin() && ! class_exists($name) && ! interface_exists($name)) {
+            if (null !== $parameter) {
+                throw UnexpectedValueException::invalidParameterTypeHint(
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection/DefaultPersistentCollectionGenerator.php]: This should not be a `Proxy\UnexpectedValueException`. Either create a new exception class or throw a `PersistentCollectionException`.

**Comment by alcaeus** [comment]
@malarzm I was going to ask for some tests, then I saw that we currently don't have any. Would it be worth to create some tests for this class?

**Comment by malarzm** [comment]
IIRC I wasn't originally writing tests for this specific class because for any of custom collection test this generator must be well and alive :)

**Comment by Maks3w** [comment]
This test class cover the scenario for PHP <7 but it's heavily coupled to the whole MongoDB setup and I don't have time for to perform a refactor and decouple the tests.

Sorry for not to be more useful but don't expect more changes from my side. The branch is open so any of you can add more commits

**Comment by malarzm** [comment]
> Sorry for not to be more useful but don't expect more changes from my side.

@Maks3w mind changing at least the exception to `PersistentCollectionException` so we could merge this as is and take care of tests on our own later?

**Comment by malarzm** [comment]
Closing in favour of #1543. Thanks for getting us started on this @Maks3w!

**Pull request closed without merging**